src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+﻿using System.Linq;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System;
+using D2L.CodeStyle.Analyzers.Extensions;
+using System.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Async {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class BlockingAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Diagnostics.AsyncMethodCannotBeBlocking,
+			Diagnostics.AsyncMethodCannotCallBlockingMethod,
+			Diagnostics.OnlyCallBlockingMethodsFromMethods,
+			Diagnostics.BlockingCallersMustBeBlocking,
+			Diagnostics.UnnecessaryBlocking,
+			Diagnostics.DontIntroduceBlockingInImplementation,
+			Diagnostics.NonBlockingImplementationOfBlockingThing
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.ConfigureGeneratedCodeAnalysis(
+				GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics
+			);
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		private static void CompilationStart( CompilationStartAnalysisContext ctx ) {
+			var attr = ctx.Compilation.GetTypeByMetadataName(
+				"D2L.CodeStyle.Annotations.BlockingAttribute"
+			);
+
+			if( attr == null ) {
+				return;
+			}
+
+			var havePossiblyUnusedBlockingAttribute = new ConcurrentDictionary<IMethodSymbol, bool>(

```

Comment:
I've been a fan of target-type new. Just sharing, not asking you to change it if you prefer the var.